### DISCUSSION TITLE
Compile Tinker9 on Bridges2

### DISCUSSION THREAD

**Discussion by blakemertz** [TinkerTools/tinker-gpu#175] [General]
I'm trying to compile T9 (CPU and GPU) on Bridges2 (they have V100 GPUs). Below is my environment setup:

- Compilers: gfortran/gcc/g++ v. 8.3.1, clang v. 9.0.1, cmake v. 3.20 (had to clone from git and install locally, Bridges2 only has v. 3.11)
- Modules loaded: cuda/11.1.1, nvhpc/21.7, fftw/3.3.8
- Had to set CUDACXX to /jet/packages/nvidia/hpc_sdk/21.7/Linux_x86_64/21.7/cuda/11.4/bin/nvcc

After cloning tinker9 from git and running git submodule update --init to update the fortran source files, I created separate build sub-directories for the cpu and gpu codes (cpu-t9 and gpu-t9 respectively).

When I ran cmake to compile the CPU-only version of T9: "cmake -DHOST=ON ..", my compile made it to 74% and then failed. Below is my error:
[bmertz@bridges2-login012 cpu-t9]$ make -j
[  0%] Building CXX object ext/interface/CMakeFiles/tinkerObjCpp.dir/cpp/tinker/routines.cpp.o
[  0%] Building CXX object ext/interface/CMakeFiles/tinkerObjCpp.dir/cpp/tinker/modcpp.cpp.o
[  0%] Built target __t9_copy_shell_exe
"/ocean/projects/bio200023p/bmertz/binaries/tinker9/ext/interface/cpp/tinker/routines.cpp", line 1: catastrophic error: cannot open source file "macro.hh"
  #include "macro.hh"
                     ^

1 catastrophic error detected in the compilation of "/ocean/projects/bio200023p/bmertz/binaries/tinker9/ext/interface/cpp/tinker/routines.cpp".
Compilation terminated.
"/ocean/projects/bio200023p/bmertz/binaries/tinker9/ext/interface/cpp/tinker/detail/action.hh", line 3: catastrophic error: cannot open source file "macro.hh"
  #include "macro.hh"
                     ^

1 catastrophic error detected in the compilation of "/ocean/projects/bio200023p/bmertz/binaries/tinker9/ext/interface/cpp/tinker/modcpp.cpp".
Compilation terminated.
make[2]: *** [ext/interface/CMakeFiles/tinkerObjCpp.dir/build.make:75: ext/interface/CMakeFiles/tinkerObjCpp.dir/cpp/tinker/routines.cpp.o] Error 2
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [ext/interface/CMakeFiles/tinkerObjCpp.dir/build.make:88: ext/interface/CMakeFiles/tinkerObjCpp.dir/cpp/tinker/modcpp.cpp.o] Error 2
make[1]: *** [CMakeFiles/Makefile2:461: ext/interface/CMakeFiles/tinkerObjCpp.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 74%] Built target tinkerObjF
make: *** [Makefile:91: all] Error 2

It appears that macro.hh is missing in the cpp portion of the directory tree (it is also located in the ifort and gfortran portions of the tree). Should I be pointing the compiler to a different location for macro.hh or generate a macro.hh file? Thanks.

ADDITIONAL NOTE: I get the same error when trying to compile the GPU version of the code after running "cmake -DCOMPUTE_CAPABILITY=70 .."

**Comment by zhi-wang** [comment]
The error you saw about the missing "macro.hh" file is because the default Fortran compiler is set to "pgf95" (under the nvhpc installation) for some reason. This compiler has a long-lasting bug so that it cannot be used for Tinker9. Please set the environmental variable "FC=your_gfortran" to overwrite the default setting.

Another thing is, CUDACXX is supposed to be set to the `nvcc` under the `cuda/11.1.1` installation, not the one under the `nvhpc`. However, this may not be a problem for cmake 3.20. This is not allowed for older cmake versions because they don't have support for the nvhpc compilers. If there is no problem with 3.20, I would not worry about it.

**Comment by zhi-wang** [comment]
An error message is added to help diagnose this issue. (In commit 77144844e40deb68fad0d1d24602aa3f54a6bd48)

**Comment by blakemertz** [comment]
Zhi,

Thanks for the tips, finally got tinker9 GPU to compile. However, when I went to run the tests, I get the following error:

/ocean/projects/bio200023p/bmertz/binaries/tinker9/src/test/test_info.cpp:8: FAILED:
due to unexpected exception with message:
  Errno 35 (CUDA driver version is insufficient for CUDA runtime version) at /
  ocean/projects/bio200023p/bmertz/binaries/tinker9/src/cudart/gpu_card.cpp:288

===============================================================================
test cases: 1 | 1 failed
assertions: 1 | 1 failed

make[3]: *** [CMakeFiles/test.dir/build.make:70: CMakeFiles/test] Error 1
make[2]: *** [CMakeFiles/Makefile2:403: CMakeFiles/test.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:410: CMakeFiles/test.dir/rule] Error 2
make: *** [Makefile:215: test] Error 2

Is this simply because I'm running on the login node instead of submitting to a GPU node? If so, I'll submit to the GPU queue. Just wanted to make sure I wasn't missing anything. Thanks again for all your help today.

Blake


Helen C. Levitt Visiting Professor

Carver College of Medicine

University of Iowa

-----------------------------------------------------------

Associate Professor

C. Eugene Bennett Department of Chemistry

WVU Cancer Institute

Director of Research Computing

(304) 293-9166
www.mertzlab-biophysics.com<http://www.mertzlab-biophysics.com/index.html>

https://docs.hpc.wvu.edu/

"Life is not easy for any of us. But what of that? We must have perseverance and above all confidence in ourselves. We must believe that we are gifted for something and that this thing must be attained." Marie Curie

"Start by doing what's necessary; then do what's possible; and suddenly you are doing the impossible." St. Francis of Assisi

________________________________

**Reply by zhi-wang** [reply]
This happens when you are trying to use a newer cuda runtime but the cuda driver installed on the machine is older. The system admin should be able to fix it.

**Reply by zhi-wang** [reply]
You either use an older runtime or install a newer driver.

**Comment by blakemertz** [comment]
Quick update -- I'm running Tinker in interactive mode on Bridges2 GPU. I'm completely befuddled by the header of my log file though, as the code declares itself to be tinker version 8.10.1:
     ######################################################################
   ##########################################################################
  ###                                                                      ###
 ###            Tinker  ---  Software Tools for Molecular Design            ###
 ##                                                                          ##
 ##                       Version 8.10.1  October 2021                       ##
 ##                                                                          ##
 ##               Copyright (c)  Jay William Ponder  1990-2021               ##
 ###                           All Rights Reserved                          ###
  ###                                                                      ###
   ##########################################################################
     ######################################################################

 Primary GPU package :  CUDA

 GPU Device :  Setting Device ID to 0 from GPU utilization
 Molecular Dynamics Trajectory via Nose-Hoover NPT Algorithm

This was the command I executed on my lipid system:
[bmertz@v014 dmpc128-4]$ /ocean/projects/bio200023p/bmertz/binaries/tinker9/gpu-t9/dynamic9.sh dmpc128-4-npt-s1.xyz -k dmpc128-4-npt-s1.key 500000 1.0 10.0 4 310 1.0 1000 > dmpc128-4-npt-s1.log

Is this simply an error or am I actually running tinker8? Because I cloned the source code from Tinker9. And my simulations I've been running on Argon at U. of Iowa have Tinker9 in the header of all of my log files. Thanks.

**Reply by zhi-wang** [reply]
If it shows "cuda platform" then it is definitely Tinker9. The suspicious Tinker 8 output should be a recent bug. I used to hijack the printing in the initialization of the dependent Fortran Tinker routines, and I recently cleaned up some code interfacing with them. Obviously the new output for Tinker9 was not printed. Thank you for this report!

**Reply by zhi-wang** [reply]
fixed in 44806f68fa297eac08bfcfc0f458fce357bdb5ae
